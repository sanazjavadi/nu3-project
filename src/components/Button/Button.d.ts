declare namespace IButton {
  export interface IProps {
    id: string;
    type?: string;
    handleClick: () => void;
  }
}

export { IButton };
