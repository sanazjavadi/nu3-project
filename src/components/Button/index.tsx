import "./index.css";
import useCard from "../../hooks/useCard";
import { FaMinus, FaPlus, FaTrash } from "react-icons/fa";

type AmountButtonsProps = {
  id: string;

  handleClick: () => void;
};

const Button = ({ id, handleClick }: AmountButtonsProps) => {
  const { updateItemQty, getItemQty } = useCard();
  const amount = getItemQty(id);

  return getItemQty(id) > 0 ? (
    <div className="amount-buttons-wrapper amount-btns">
      <button
        type="button"
        className="amount-btn"
        onClick={() => updateItemQty(id, -1)}
      >
        {amount > 1 ? <FaMinus /> : <FaTrash />}
      </button>
      <h2 className="amount">{amount}</h2>
      <button
        type="button"
        className="amount-btn"
        onClick={() => updateItemQty(id, 1)}
      >
        <FaPlus />
      </button>
    </div>
  ) : (
    <button
      onClick={handleClick}
      className="inline-block px-6 py-2.5 bg-gray-800 text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-gray-900 hover:shadow-lg focus:bg-gray-900 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-gray-900 active:shadow-lg transition duration-150 ease-in-out"
    >
      Add to cart
    </button>
  );
};

export default Button;
