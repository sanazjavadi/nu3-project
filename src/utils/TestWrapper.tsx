import { configureStore, Store } from "@reduxjs/toolkit";
import { MemoryRouter } from "react-router-dom";
import { ComponentType, FunctionComponent } from "react";
import { Provider } from "react-redux";
import reducer from "../reducers";

export const makeStore = (): Store => {
  return configureStore({ reducer });
};

const wrapComponent = (
  Component: ComponentType | FunctionComponent | any,
  store: Store | null = null,
  props = {}
): JSX.Element => {
  return (
    <Provider store={store || makeStore()}>
      <MemoryRouter>
        <Component {...props} />
      </MemoryRouter>
    </Provider>
  );
};

export default wrapComponent;
