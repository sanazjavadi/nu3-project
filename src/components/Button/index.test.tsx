import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import TestWrapper, { makeStore } from "../../utils/TestWrapper";
import Button from "./";

const store = makeStore();

const MockData = [
  {
    price: 13.99,
    image:
      "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/4_252F5_252F3_252F4_252F453437c2922e342acc8db2a47e2a24ee197fe940_20174_9807111a-5f4b-4430-a21b-3809799931aa_large.jpg?v=1623047667",
    vendor: "BEAVITA",
    title: "BEAVITA Vitalkost DiÃ¤t-Shake",
    id: "4365140131921",
    link: "/products/beavita-vitalkost-diaet-shake",
  },
];

describe("Button Tests Cases", () => {
  test("Button: Render Button with More Than 1 Qty", () => {
    render(TestWrapper(Button, store, { id: MockData[0].id }));
    const initialButtonElement = screen.getByTestId("initial-button");

    expect(initialButtonElement).toBeInTheDocument();
  });

  test("Button: onClick should be called", () => {
    const onClick = jest.fn();
    render(
      TestWrapper(Button, store, {
        id: MockData[0].id,
        handleClick: onClick,
      })
    );

    const initialButtonElement = screen.getByTestId("initial-button");
    fireEvent.click(initialButtonElement);
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});

test("Button: Snapshot", () => {
  const component = renderer.create(
    TestWrapper(Button, store, { id: MockData[0].id })
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
