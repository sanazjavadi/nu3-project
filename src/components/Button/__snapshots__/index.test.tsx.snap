// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button: Snapshot 1`] = `
<button
  className="inline-block px-6 py-2.5 bg-gray-800 text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-gray-900 hover:shadow-lg focus:bg-gray-900 focus:shadow-lg focus:outline-none focus:ring-0 active:bg-gray-900 active:shadow-lg transition duration-150 ease-in-out"
  data-testid="initial-button"
>
  Add to cart
</button>
`;
